File: dsa/MergeSort.java
class MergeSort{
    public static void conquer(int arr[], int si, int mid, int ei) {
        int merged[] = new int[ei - si + 1];
        int idx1 = si, idx2 = mid+1, x = 0;

        while(idx1 <= mid && idx2 <= ei) {
            if(arr[idx1] <= arr[idx2]) {
                merged[x++] = arr[idx1++];
            } else {
                merged[x++] = arr[idx2++];
            }
        }

        while(idx1 <= mid) {
            merged[x++] = arr[idx1++];
        }

        while(idx2 <= ei) {
            merged[x++] = arr[idx2++];
        }

        for(int i = si, j = 0; j < merged.length; i++, j++) arr[i] = merged[j]; 
    }

    public static void divide(int arr[], int si, int ei) {
        if(si >= ei) return;
        
        int mid  = si + (ei - si)/2;
        divide(arr, si, mid);
        divide(arr, mid+1, ei);

        conquer(arr, si, mid, ei);
    }


    public static void main(String args[]){
        int arr[] = {1,3, 4, 2, 6, 3, 9};
        divide(arr, 0,arr.length-1);
        for(int i : arr) System.out.print(i + " ");
        System.out.println();
    }
}

File: dsa/mergeSort.cpp
#include <iostream>
using namespace std;

void merge(int *arr, int s, int e)
{
    int mid = s + (e - s) / 2;
    int len1 = mid - s + 1;
    int len2 = e - mid;

    int *first = new int[len1];
    int *second = new int[len2];

    int k = s;

    for (int i = 0; i < len1; i++)
    {
        first[i] = arr[k++];
    }

    for (int i = 0; i < len2; i++)
    {
        second[i] = arr[k++];
    }

    int i1 = 0, i2 = 0;
    k = s;

    while (i1 < len1 && i2 < len2)
    {
        if (first[i1] <= second[i2])
        {
            arr[k++] = first[i1++];
        }
        else
            arr[k++] = second[i2++];
    }

    while (i1 < len1)
    {
        arr[k++] = first[i1++];
    }

    while (i2 < len2)
    {
        arr[k++] = second[i2++];
    }

    delete[] first;
    delete[] second;
}

void mergeSort(int *arr, int s, int e)
{
    if (s >= e)
        return;

    int mid = s + (e - s) / 2;

    mergeSort(arr, s, mid);
    mergeSort(arr, mid + 1, e);

    merge(arr, s, e);
}

int main()
{
    int arr[] = {4, 2, 1, 3, 5, 9, 6};
    int n = sizeof(arr) / sizeof(arr[0]);

    mergeSort(arr, 0, n - 1);

    for (int i : arr)
        cout << i << " ";

    return 0;
}